runtime/runtime/src/genesis.rs

```diff
@@ -71,210 +69,306 @@ impl<'a> StorageComputer<'a> {
     }
 
     /// Adds multiple StateRecords to the users' storage info.
-    pub fn process_records(&mut self, records: &[StateRecord]) {
+    pub(crate) fn process_records(&mut self, records: &[StateRecord]) {
         for record in records {
             self.process_record(record);
         }
     }
 
     /// Returns the current storage use for each user.
-    pub fn finalize(self) -> HashMap<AccountId, u64> {
+    pub(crate) fn finalize(self) -> HashMap<AccountId, u64> {
         self.result
     }
 }
 
-pub struct GenesisStateApplier {}
+const MAX_OUTSTANDING_WRITES: usize = 0x100_000;
+type State = (usize, StateRoot, TrieUpdate);
 
-impl GenesisStateApplier {
-    fn commit(
-        mut state_update: TrieUpdate,
-        current_state_root: &mut StateRoot,
-        tries: &mut ShardTries,
-        shard_uid: ShardUId,
-    ) {
+pub(crate) struct AutoFlushingTrieUpdate<'a> {
+    active_writers: &'a atomic::AtomicUsize,
+    tries: &'a ShardTries,
+    shard_uid: ShardUId,
+    state: Mutex<State>,
+}
+
+impl<'a> AutoFlushingTrieUpdate<'a> {
+    pub(crate) fn new(
+        active_writers: &'a atomic::AtomicUsize,
+        state_root: StateRoot,
+        tries: &'a ShardTries,
+        uid: ShardUId,
+    ) -> Self {
+        let state = Mutex::new((0, state_root, tries.new_trie_update(uid, state_root)));
+        Self { active_writers, state, tries, shard_uid: uid }
+    }
+
+    fn lock(&self) -> std::sync::MutexGuard<State> {
+        self.state.lock().unwrap_or_else(|g| g.into_inner())
+    }
+
+    fn modify<R>(&self, cb: impl FnOnce(&mut StateRoot, &mut TrieUpdate) -> R) -> R {
+        let mut guard = self.lock();
+        let (ref mut update_budget, ref mut state_root, ref mut state_update) = &mut *guard;
+        // Try to get some budget from the global pool.
+        if *update_budget == 0 {
+            let writers = self.active_writers.fetch_add(1, atomic::Ordering::SeqCst) + 1;
+            *update_budget = MAX_OUTSTANDING_WRITES / writers;
```

Comment:
I don't get how this write-budgeting works.

I looks to me like `update_budget == 0` detects when we are the first write operation to be enqueued in our `AutoFlushingTrieUpdate` instance. If so, we lazily allocate some budget from an implicit pool that is shared with all other `AutoFlushingTrieUpdate` instance sharing the same `active_writers` counter. All subsequent writes on this object will use up one of the allocated writes. When the update_budget reaches zero, we flush and remove ourselves from the active writers, which kind of returns the outstanding writes to the pool.

First question: Is my understanding right?


Second question: Does this really limit the total number of outstanding writes? It seems like the first writer gets `MAX_OUTSTANDING_WRITES` / 2, the next gets `MAX_OUTSTANDING_WRITES` / 3, the next `MAX_OUTSTANDING_WRITES` / 4 and so on. (A diverging series, so unlimited outstanding writes as the number of writer grows.)

Third question: This seems a bit complicated... Why not, let's say give a fixed number to each `AutoFlushingTrieUpdate`?

Reply:
> First question: Is my understanding right?

Yes, the understanding here is correct.

> Second question

The upper limit is indeed not `MAX_OUTSTANDING_WRITES`, but the number of outstanding writers is bounded by the number of shards processed in parallel. This should probably be named `TARGET_OUTSTANDING_WRITES`, and a comment what this actually means would help too, sure.

> Third question

As the number of shards increases, a constant limit for `AutoFlushingTrieUpdate` would mean a proportional increase in the memory usage. We could set the limit appropriate for, say, 100 shards, today, but that would mean excessive flushes today, as well as on the long tail when the last shard is being processed. Admittedly I haven’t measured the impact of flushing frequency (due to huge iteration times), so I went with the intuition that keeping it as infrequent as possible is ideal.

Reply:
Ok, I see, thanks for elaborating.

It's a bit weird to me that we use some arbitrary number to limit the consumption and then go above that nuber with already 3 concurrent `AutoFlushingTrieUpdate` objects. But I did the math quickly and it seems even with 100 concurrent readers we only use 4x the target size. I suppose that limit actually works out pretty nicely when looking into it this way.

Other than that, it makes sense to me now.

Oh and I like the removal of the Mutex :+1: